File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/netty/EventSender.java
package ru.kmorozov.library.data.loader.netty;

import io.netty.bootstrap.Bootstrap;
import io.netty.channel.Channel;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioSocketChannel;
import ru.kmorozov.gbd.logger.Logger;

import java.io.Closeable;

public final class EventSender implements Closeable {

    private static final Logger logger = Logger.getLogger(EventSender.class);

    private final String server;
    private final int port;
    private EventLoopGroup group;
    private Channel channel;
    private boolean isInitCompleted;

    private boolean initIsNotPossible;

    public static final EventSender INSTANCE = new EventSender("localhost", 5252);

    private EventSender(String server, int port) {
        this.server = server;
        this.port = port;
    }

    private void init() {
        this.group = new NioEventLoopGroup();
        Bootstrap bootstrap = new Bootstrap().group(this.group)
                .channel(NioSocketChannel.class)
                .handler(new ClientAdapterInitializer());
        try {
            this.channel = bootstrap.connect(this.server, this.port).sync().channel();

            this.isInitCompleted = true;
        } catch (InterruptedException e) {
            this.group.shutdownGracefully();

            EventSender.logger.error(e);
        } catch (final Exception ex) {
            EventSender.logger.error(ex.getMessage());
            this.initIsNotPossible = true;
        }
    }

    private void write(String msg) {
        if (!this.isInitCompleted)
            if (this.initIsNotPossible)
                return;
            else
                this.init();

        if (this.isInitCompleted) {
            this.channel.write(msg);
            this.channel.flush();
        }
    }

    @Override
    public void close() {
        this.group.shutdownGracefully();
    }

    public void sendInfo(Logger logger, String msg) {
        try {
            this.write(msg);
        } catch (Exception ex) {
            logger.error(ex.getMessage());
        }

        logger.info(msg);
    }
}

File: ru.kmorozov.library.data.server/src/main/java/ru/kmorozov/library/data/loader/netty/EventSender.kt
package ru.kmorozov.library.data.loader.netty

import io.netty.bootstrap.Bootstrap
import io.netty.channel.Channel
import io.netty.channel.EventLoopGroup
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioSocketChannel
import ru.kmorozov.gbd.logger.Logger

import java.io.Closeable

class EventSender private constructor(private val server: String, private val port: Int) : Closeable {
    private var group: EventLoopGroup? = null
    private var channel: Channel? = null
    private var isInitCompleted: Boolean = false

    private var initIsNotPossible: Boolean = false

    private fun init() {
        group = NioEventLoopGroup()
        val bootstrap = Bootstrap().group(group!!)
                .channel(NioSocketChannel::class.java)
                .handler(ClientAdapterInitializer())
        try {
            channel = bootstrap.connect(server, port).sync().channel()

            isInitCompleted = true
        } catch (e: InterruptedException) {
            group!!.shutdownGracefully()

            logger.error(e)
        } catch (ex: Exception) {
            logger.error(ex.message!!)
            initIsNotPossible = true
        }

    }

    private fun write(msg: String) {
        if (!isInitCompleted)
            if (initIsNotPossible)
                return
            else
                init()

        if (isInitCompleted) {
            channel!!.write(msg)
            channel!!.flush()
        }
    }

    override fun close() {
        group!!.shutdownGracefully()
    }

    fun sendInfo(logger: Logger, msg: String) {
        try {
            write(msg)
        } catch (ex: Exception) {
            logger.error(ex.message!!)
        }

        logger.info(msg)
    }

    companion object {

        private val logger = Logger.getLogger(EventSender::class.java)

        val INSTANCE = EventSender("localhost", 5252)
    }
}